### PR TITLE
[Merged by Bors] - feat(AlgebraicGeometry/EllipticCurve/Jacobian): implement group law for Jacobian coordinates

### PR DESCRIPTION
**Pull request by Multramate**
Completes the proof of the group law in Jacobian coordinates, analogously to #8485 

---

- [x] depends on: #9432 
- [x] depends on: #13060 
- [x] depends on: #9433 
- [x] depends on: #9435 
- [x] depends on: #9436 

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by alexjbest** [comment]
So do you want both types of coordinates in Mathlib? What is the motivation for doing it that way, rather than just choosing one type of coordinates?

**Comment by Multramate** [comment]
> So do you want both types of coordinates in Mathlib? What is the motivation for doing it that way, rather than just choosing one type of coordinates?

I need Jacobian coordinates, because their coordinates are precisely the division polynomials in #6703 and the phi_n/omega_n defined in AEC Exercise III.3.7, [albeit with an error](https://math.stackexchange.com/questions/4675977/division-polynomials-in-characteristic-two). I don't need projective coordinates at the moment, but I believe it might be a very useful definition in general (e.g. to give an elementary definition of reduction modulo p without any bias on the weights). In fact projective coordinates was my toy model for Jacobian coordinates because originally I wasn't sure how to write the API at all.

**Comment by alreadydone** [comment]
I'd like to merge master into this soon, because I'm building on top of this branch but `WeierstrassCurve.map` is not available here. I expect this PR to be merged as a whole so I won't update the "prerequisite" PRs. Before I merge I'd like to get #12874 in.

**Comment by Multramate** [comment]
Sure, give me a couple of hours and I'll hopefully update everything by today!

**Comment by alreadydone** [comment]
[Here](https://github.com/leanprover-community/mathlib4/compare/EllipticCurve.Jacobian...EllipticCurve.Jacobian+?expand=1)'s some additions and changes I made to the Jacobian file for the purpose of division polynomials, and maybe you want to incorporate them into one or more of your PRs. I still think it makes sense to merge this PR as a whole and close the others.

**Comment by Multramate** [comment]
Thanks! I'll have a look later in the week. I don't mind merging this PR as a whole, but I don't want to set a precedent where 1000+ line feat PRs are acceptable. Maybe we can do this for `Projective` once the `Jacobian` ones are done.

**Comment by Multramate** [comment]
@alreadydone do we actually need these?
```
lemma addZ_self {P : Fin 3 → R} : addZ P P = 0
lemma addX_self {P : Fin 3 → R} (hP : V.Equation P) : V.addX P P = 0
lemma addY'_self {P : Fin 3 → R} : V.addY' P P = 0
lemma addY_self {P : Fin 3 → R} (hP : V.Equation P) : V.addY P P = 0
```
Looks to me that they should never be accessed if we're using `add`.

**Comment by alreadydone** [comment]
Those addXYZ_self lemmas were mainly introduced to prove [addXYZ_smulField](https://github.com/leanprover-community/mathlib4/blob/DivisionPolynomial_smul/Mathlib/AlgebraicGeometry/EllipticCurve/DivisionPolynomial.lean#L565-L567) for all m,n. It's equally valid in the m=n case, and although it's just a lemma towards the main theorem it could be useful elsewhere.

The generalization of scalar multiplication on `Fin 3 → R` from `R` to `Rˣ` is for the same purpose; it would be ugly to use `Units.mk0`, which still won't cover the m=n case. Although the case n=m+1 suffices for the main theorem, it's nicer to state and prove the general case.

File: Mathlib/AlgebraicGeometry/EllipticCurve/Jacobian.lean
Comment (Multramate) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Jacobian.lean]: Sorry I missed this.

File: Mathlib/AlgebraicGeometry/EllipticCurve/Group.lean
Comment (alreadydone) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Group.lean:677]: I think you can just inline all the lemmas and remove the namespace section:
```suggestion
noncomputable instance WeierstrassCurve.Jacobian.instAddCommGroupPoint
    {F : Type u} [Field F] {W : Jacobian F} : AddCommGroup W.Point where
```
I don't think anyone would use those lemmas, and they confuse Lean if not protected.

**Review by alreadydone — APPROVED** [review]
LGTM otherwise

File: Mathlib/AlgebraicGeometry/EllipticCurve/Group.lean
Comment (alreadydone) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Group.lean:51]:
```suggestion
 * `WeierstrassCurve.Jacobian.instAddCommGroupPoint`: the type of nonsingular rational
```

File: Mathlib/AlgebraicGeometry/EllipticCurve/Group.lean
Comment (Multramate) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Group.lean:677]: Good point - I'll hide the affine ones too. I'll keep the namespace because I'll delete my instance name.

File: Mathlib/AlgebraicGeometry/EllipticCurve/Group.lean
Comment (alreadydone) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Group.lean:618]: Thanks! I have some golfs [here](https://github.com/leanprover-community/mathlib4/commit/09ad43d968d35eb6069d426d3c6e882ba31702cb).
Using [Function.Injective.addCommGroup](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Group/InjSurj.html#Function.Injective.addCommGroup) might be able to save us some lines but you'd still need to introduce nsmul, zsmul, and sub with three `letI`s.

**Comment by alreadydone** [comment]
Thanks and congratulations 🎉 
maintainer merge

File: Mathlib/AlgebraicGeometry/EllipticCurve/Group.lean
Comment (Multramate) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Group.lean:618]: Thanks! I would hope that a friendlier version of [Function.Injective.addCommGroup](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Group/InjSurj.html#Function.Injective.addCommGroup) exists. Surely for any reasonable structure we can deduce `nsmul`, `zsmul`, and `sub` automatically from the others...?

Comment (alreadydone) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Group.lean:618]: But you may not want to use the affine operations in this case. toAffine need to check whether the z coordinate is zero, which is not decidable. 

In Function.Injective.addCommGroup the SMul and Sub instances are supplied to avoid potential diamonds, which is probably not a concern here.

**Comment by riccardobrasca** [comment]
Thanks!!!

bors merge

**Pull request closed without merging**
